pennylane/devices/default_qubit_torch.py

```diff
@@ -0,0 +1,324 @@
+    _asnumpy = staticmethod(lambda x: x.cpu().numpy())
+    _conj = staticmethod(torch.conj)
+    _imag = staticmethod(torch.imag)
+    _norm = staticmethod(torch.norm)
+    _flatten = staticmethod(torch.flatten)
+
+    def __init__(self, wires, *, shots=None, analytic=None, torch_device="cpu"):
+        self._torch_device = torch_device
+        super().__init__(wires, shots=shots, cache=0, analytic=analytic)
+
+        # Move state to torch device (e.g. CPU, GPU, XLA, ...)
+        self._state.requires_grad = True
+        self._state = self._state.to(self._torch_device)
+        self._pre_rotated_state = self._state
+
+    @staticmethod
+    def _asarray(a, dtype=None):
+        if isinstance(a, list):
+            if not isinstance(a[0], torch.Tensor):
+                res = np.asarray(a)
+            else:
+                res = torch.cat([torch.reshape(i, (-1,)) for i in a], dim=0)
+        else:
+            res = torch.as_tensor(a, dtype=dtype)
+        return res
+
+    @staticmethod
+    def _dot(x, y):
+        if x.device != y.device:
+            if x.device != "cpu":
+                return torch.tensordot(x, y.to(x.device), dims=1)
+            if y.device != "cpu":
+                return torch.tensordot(x.to(y.device), y, dims=1)
+
+        return torch.tensordot(x, y, dims=1)
+
+    def _cast(self, a, dtype=None):
+        return torch.as_tensor(self._asarray(a, dtype=dtype), device=self._torch_device)
+
+    @staticmethod
+    def _reduce_sum(array, axes):
+        if not axes:
+            return array
+        return torch.sum(array, dim=axes)
+
+    @staticmethod
+    def _conj(array):
+        if isinstance(array, torch.Tensor):
+            return torch.conj(array)
+        return np.conj(array)
+
+    @staticmethod
+    def _ravel_multi_index(multi_index, dims):
+        # Idea: ravelling a multi-index can be expressed as a matrix-vector product
+        flip = lambda x: torch.flip(x, dims=[0])
+        dims = torch.as_tensor(dims, device=multi_index.device)
+        coeffs = torch.ones_like(dims, device=multi_index.device)
+        coeffs[:-1] = dims[1:]
+        coeffs = flip(torch.cumprod(flip(coeffs), dim=0))
+        ravelled_indices = (multi_index.T.type(torch.float) @ coeffs.type(torch.float)).type(
+            torch.long
+        )
+        return ravelled_indices
+
+    @staticmethod
+    def _scatter(indices, array, new_dimensions):
+
+        # `array` is now a torch tensor
+        tensor = array
+        new_tensor = torch.zeros(new_dimensions, dtype=tensor.dtype, device=tensor.device)
+        new_tensor[indices] = tensor
+        return new_tensor
+
+    @staticmethod
+    def _allclose(a, b, atol=1e-08):
+        return torch.allclose(a, torch.as_tensor(b, dtype=a.dtype), atol=atol)
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(passthru_interface="torch", supports_reversible_diff=False)
```

Comment:
So the `capabilities` method does get tested, so I don't know what's up with code cov here.

Reply:
@albi3ro we discovered yesterday that any instance of `pass` in the code gets ignored by codecov due to a line in the `.coveragerc` file. #1555 should fix this :tada: